Ngcc perf improvements

petebacondarwin opened the pull request:
This PR sits on top of #30591

petebacondarwin committed:
fix(ivy): ngcc - remove unwanted logging message

This message gets called if a format has already been
compiled and we only want the first. So the message itself
is wrong but it is also not very useful anyway.

petebacondarwin committed:
perf(ivy): ngcc - add a cache to the FileSystem

When profiling ngcc it is notable that a large amount of time
is spent dealing with an exception that is thrown (and handled
internally by fs) when checking the existence of a file.

We check file existence a lot in both finding entry-points
and when TS is compiling code. This commit adds a simple
cached `FileSystem`, which wraps a real `FileSystem` delegate.
This will reduce the number of calls through to `fs.exists()` and
`fs.readFile()` on the delegate.

Initial benchmarks indicate that the cache is miss to hit ratio
for `exists()` is about 2:1, which means that we save about 1/3
of the calls to `fs.existsSync()`.

Note that this implements a "non-expiring" cache, so it is not suitable
for a long lived `FileSystem`, where files may be modified externally.
The cache will be updated if a file is changed or moved via
calls to `FileSystem` methods but it will not be aware of changes
to the files system from outside the `FileSystem` service.

For ngcc we must create a new `FileSystem` service
for each run of `mainNgcc` and ensure that all file operations
(including TS compilation) use the `FileSystem` service.
This ensures that it is very unlikely that a file will change
externally during `mainNgcc` processing.

petebacondarwin committed:
perf(ivy): ngcc - only find dependencies when targeting a single entry-point

Previously, ngcc had to walk the entire `node_modules` tree looking for
entry-points, even if it only needed to process a single target entry-point
and its dependencies.

This added up to a few seconds to each execution of ngcc, which is noticeable
when being run via the CLI integration.

Now, if an entry-point target is provided, only that target and its entry-points
are considered rather than the whole folder tree.

alan-agius4 commented on packages/compiler-cli/ngcc/src/entry_point_finder/utils.ts:
```diff
@@ -0,0 +1,75 @@
+ * Extract all the values (not keys) from an object.
+ * @param obj The object to process.
+ */
+function values<T>(obj: {[key: string]: T}): T[] {
```
Why not use `Object.values` directly?

petebacondarwin replied:
I didn't actually know about that function and it didn't come up in intellisense because we were using `"lib": ["es2015"]` in our tsconfig. Updating to es2018 opens this up to us!

petebacondarwin replied:
Fixed - thanks

gkalpak commented on packages/compiler-cli/ngcc/src/entry_point_finder/targeted_entry_point_finder.ts:
```diff
@@ -0,0 +1,101 @@
+    // If we get here then none of the `basePaths` matched the `entryPointPath`, which
+    // is somewhat unexpected and means that this entry-point lives completely outside
+    // any of the `basePaths`.
+    // All we can do is assume that his entry-point is a primary entry-point to a package.
```
Shouldn't we check that there is an `${entryPointPath}/package.json`?

petebacondarwin replied:
I'm not too worries as this would get picked up later when we try to actually create entry-points from these paths.

gkalpak commented on packages/compiler-cli/ngcc/src/entry_point_finder/targeted_entry_point_finder.ts:
```diff
@@ -0,0 +1,101 @@
+            return packagePath;
+          }
+        }
+      }
```
Since it is guaranteed that no `basePath` will be a subdirectory of another `basePath`, there can be max. one `basePath` matching `entryPointPath` (i.e. `entryPointPath` can only start with max. one `basePath`). 

So, if we have entered the `if (entryPointPath.startsWith(basePath.toString()))` block and still didn't find a `packagePath`, we can `break` out of the `for (const basePath of this.basePaths)` loop.

gkalpak commented on packages/compiler-cli/ngcc/src/entry_point_finder/targeted_entry_point_finder.ts:
```diff
@@ -0,0 +1,101 @@
+        // Skip up to the last nested node_modules folder if the relative entryPointPath
+        // contains any.
+        let nodeModulesIndex = segments.lastIndexOf('node_modules');
+        while (nodeModulesIndex > 0) {
```
Did you mean `>= 0`? As it is, the `segments` array will have `node_modules` as its first element after the `while` loop.
From what I can tell by the comments below, the intention was to have `node_modules` be part of `packagePath` after the `while` loop.

petebacondarwin replied:
I also just realised that to be entirely FS independent you can't just `split('/')` really!

gkalpak replied:
But it is a `PathSegment`, which means that separators will be normalized to `/`, right?

petebacondarwin replied:
PathSegments and AbsoluteFilePaths don't have to have forward slashes. It happens that at the moment we do normalize slashes to forward slashes but that could change, and I would rather avoid any code outside the `file_system` folder pretending it knows what the internals of a path should look like.

gkalpak commented on packages/compiler-cli/ngcc/src/entry_point_finder/targeted_entry_point_finder.ts:
```diff
@@ -0,0 +1,101 @@
+        while (nodeModulesIndex > 0) {
+          packagePath = join(packagePath, segments.shift() !);
+          nodeModulesIndex--;
+        }
```
If you don't care about readability, the above few lines can be golfed into: :grin:

```ts
const segments = relative(basePath, entryPointPath).split('/');
const packagePath = join(basePath, ...segments.splice(0, segments.lastIndexOf('node_modules') + 1));
```

petebacondarwin replied:
Very cool but I think that the mutating `splice()` would cause enough confusion on a quick glance to make it less useful.

gkalpak reviewed (approved):
Nice!!! 🏇
Left some questions and minor comments, but otherwise lgtm 💯 

Commit message typos:
-  _"the **cache is** miss to hit ratio [...] is"_ --> _"the **cache's** miss to hit ratio [...] is"_
- _"changes to the files **system** from outside"_ --> _"changes to the files from outside"_ (?)
- _"the FileSystem service"_ --> _"the \`FileSystem\` service"_ (for consistency)

petebacondarwin commented on packages/compiler-cli/ngcc/main-ngcc.ts:
```diff
@@ -8,7 +8,7 @@
 import * as yargs from 'yargs';
 
-import {resolve, setFileSystem, NodeJSFileSystem} from '../src/ngtsc/file_system';
+import {resolve, setFileSystem, CachedNodeJSFileSystem} from '../src/ngtsc/file_system';
```
No idea. Sometimes it just doesn't seem to notice... :-/ I think that sometimes it fails to reformat something if it crashes during parsing - say because the code is syntactically broken.

gkalpak replied:
It should still cause `gulp lint` to break on CI, though 😕

petebacondarwin commented on packages/compiler-cli/src/ngtsc/file_system/src/cached_node_js_file_system.ts:
```diff
@@ -0,0 +1,72 @@
+    super.mkdir(path);
+    this.existsCache.set(path, true);
+  }
+}
```
I am pretty sure that TS will never call `stat()` without first calling `exists()` since calling `stat()` on a non-existent file/dir will throw an error.

alxhub replied:
Let's do this anyway.

Just because one consumer knows that a file exists doesn't mean all consumers will.

petebacondarwin commented on packages/compiler-cli/src/ngtsc/file_system/src/cached_node_js_file_system.ts:
```diff
@@ -0,0 +1,72 @@
+
+  symlink(target: AbsoluteFsPath, path: AbsoluteFsPath): void {
+    super.symlink(target, path);
+    this.existsCache.set(path, true);
```
What would happen if the symlinked file contents changed?  We would then need to track the change back to all the symlinks that point to it...

gkalpak replied:
Isn't that the case anyway? :grin:
Even if you've read the symlink itself and saved the content in `readFileCache`, you need to keep `readFileCache` up-to-date wrt changes to the target file.

Maybe we need a special case for symlinks (to reflect their targets' contents (but not their existense? 😱))?

petebacondarwin replied:
Hmm, good point...

petebacondarwin replied:
OK, so my plan, given that symlinked files are somewhat rare in ngcc scenarios, is to never cache a call to `readFile()` if it points to a symlink.

petebacondarwin commented on packages/compiler-cli/src/ngtsc/file_system/src/cached_node_js_file_system.ts:
```diff
@@ -0,0 +1,72 @@
+
+  copyFile(from: AbsoluteFsPath, to: AbsoluteFsPath): void {
+    super.copyFile(from, to);
+    this.existsCache.set(to, true);
```
That's an interesting idea. I had thought about it but didn't want to make a call to the file system to get the contents. But of course I forgot that we could access the cache directly.

That being said, filling the cache in cases like this could increase memory pressure in a case where we might not care for the contents. Given that these things are always a trade-off, I would rather not try to second-guess what is needed. Let's drive such optimizations from data. If we can show that adding such cache entries improves ngcc's speed then happy to add it later.

petebacondarwin commented on packages/compiler-cli/src/ngtsc/file_system/src/cached_node_js_file_system.ts:
```diff
@@ -0,0 +1,72 @@
+    super.moveFile(from, to);
+    this.existsCache.set(from, false);
+    this.readFileCache.delete(from);
+    this.existsCache.set(to, true);
```
That seems reasonable.

petebacondarwin commented on packages/compiler-cli/src/ngtsc/file_system/src/cached_node_js_file_system.ts:
```diff
@@ -0,0 +1,72 @@
+  readFile(path: AbsoluteFsPath): string {
+    if (!this.readFileCache.has(path)) {
+      try {
+        this.readFileCache.set(path, super.readFile(path));
```
No 😱!

ngcc is different to CLI in the sense that it is not long running. Each run only processes some subset of the node_module files and then the cache is deleted.

But I will take a look and see if there is significant difference.

petebacondarwin replied:
The heap being used does increase after this PR. Here is a spreadsheet that show heap usage for a standard run of `ivy-ngcc` (not via CLI) of all the Angular, CDK and Material packages.

https://docs.google.com/spreadsheets/d/1PjwfaBCQTUQrYZ-dx__GKm38CecjbWKit1ffopsKms4/edit?usp=sharing

The bottom line is that pre-perf the average heap usage was ~140Mb and post-perf it is ~350Mb. So more than double.

Whether this is a concern is up for debate. Note that in the CLI integration the memory usage from ngcc would be lower as the cache is deleted after each ngcc run; and each run is only compiling a subset of the packages.

gkalpak replied:
It's a sizeable increase. It doesn't sound like something most dev machines coulnd't handle, but I don't know about other environments (e.g. CI).

We should also look at how memory increases when used via the cli, because the cli will already be consuming memory for other stuff.

Maybe it would be worth capping the extra memory (e.g. limiting the number of cached entries (e.g. using an LRU), or ignoring files larger than a max size, or limiting the total characters cached, etc.) to ensure memory won't grow out of control. How would that affect performance?

In any case, the tooling team should have better insight on whether that might be a problem for end-users (since they've been looking into cli memory issues recently and have a better idea of what memory usage our users could get away with).

(cc @angular/tools)

petebacondarwin replied:
Running within the CLI (integration/cli-hello-world-ivy-compat) we get much closer results:

**Average heap usage**
pre-perf ~234Mb
post-perf ~242Mb

filipesilva replied:
How are you measuring average memory usage? I've been measuring it in a couple of different ways and have had different results depending on the method. 

If you're measuring it from within the process with `process.memoryUsage().heapUsed`, it will not give you the total memory used for that node process. 

Measuring it from within the node process overall is highly dependent on when it is measured, since the ngc/ngtsc/ngcc is sync. 

Average memory usage is also not as relevant as peak memory usage. Average memory might be low, but peak might still be high. Peak memory usage is what causes current OOM errors.

I have a tool in https://github.com/filipesilva/angular-cli-perf-benchmark#benchmark-package that you can use to measure peak and average usage, along with CPU as build time as well, over several iterations.

Also you mentioned that the cache is deleted. Is this done automatically? 

I also see that this code is within `/ngtsc/`. Is it used by Ivy, or will it be used by Ivy in the future? We already cache files in the CLI compiler host we provide for the AngularProgram. Is there an overlap here where we double cache?

petebacondarwin replied:
Thanks for this input @filipesilva. I'm afraid I forgot to remove the `merge` label so this has now landed. I suggest that we monitor closely any significant memory issues in the next week or two and fix as necessary. The obvious and simple revert would be to remove the "file contents" cache but to keep the "file exists" cache, the latter is the most important from a speed performance perspective anyway.

Regarding your questions:

* I was indeed using `process.memoryUsage().heapUsed` but only at the end of each package compilation. I felt this was good enough for this experiment since the change being tested was whether the size of the cache was causing issues, and the cache is effectively a singleton. I accept that it does not account for peaks in memory usage within the compilation, but I expected these to be equivalent both before and after the cache implementation.
* When I say that the cache is deleted, it is really just being made available for garbage collection as a new empty cache is created (replacing the old one) for each execution of the ngcc compilation (as triggered from the CLI [here](https://github.com/angular/angular-cli/blob/6a99e74a14ef2a5ce24eb6897f517095bf2b5d5a/packages/ngtools/webpack/src/ngcc_processor.ts#L72)).
* This cache is not used by ngtsc - it has no dynamic invalidation built-in and so would not be useful for a long running watching compiler, or any other scenario where the files are likely to change from outside the `FileSystem`. And as you say, CLI already provides cached access to files (via its custom `CompilerHost`) so there is no benefit in ngtsc adding one too.

petebacondarwin commented on packages/tsconfig.json:
```diff
@@ -19,7 +19,7 @@
     "rootDir": ".",
     "inlineSourceMap": true,
-    "lib": ["es5", "dom", "es2015.promise", "es2015.collection", "es2015.iterable", "es2015.core"],
+    "lib": ["es5", "dom", "es2015.promise", "es2015.collection", "es2015.iterable", "es2015.core", "es2017.object"],
```
I believe it is only used for intellisense. But that is why you as a fw-dev-infra as needed to review it :-P Perhaps @alexeagle has a view?

petebacondarwin commented on packages/compiler-cli/ngcc/src/entry_point_finder/utils.ts:
```diff
@@ -0,0 +1,65 @@
+ */
+function removeDeeperPaths(value: AbsoluteFsPath, index: number, array: AbsoluteFsPath[]) {
+  for (let i = 0; i < index; i++) {
+    if (value.startsWith(array[i].toString())) return false;
```
Not right now because I have backed out of changing `AbsoluteFsPath` so that it cannot be implicitly cast to a `string`. But if we bring that back then this would be needed. I'll remove for now.

petebacondarwin commented on packages/compiler-cli/ngcc/src/entry_point_finder/targeted_entry_point_finder.ts:
```diff
@@ -0,0 +1,101 @@
+   */
+  private computePackagePath(entryPointPath: AbsoluteFsPath): AbsoluteFsPath {
+    for (const basePath of this.basePaths) {
+      if (entryPointPath.startsWith(basePath.toString())) {
```
Same as https://github.com/angular/angular/pull/30525#discussion_r298800638

petebacondarwin commented on packages/compiler-cli/ngcc/test/entry_point_finder/targeted_entry_point_finder_spec.ts:
```diff
@@ -0,0 +1,198 @@
+      function dumpEntryPointPaths(
+          basePath: AbsoluteFsPath, entryPoints: EntryPoint[]): [string, string][] {
+        return entryPoints.map(
+            x => [relative(basePath, x.package).toString(), relative(basePath, x.path).toString()]);
```
Same as https://github.com/angular/angular/pull/30525#discussion_r298800638

petebacondarwin commented on packages/compiler-cli/ngcc/test/entry_point_finder/targeted_entry_point_finder_spec.ts:
```diff
@@ -0,0 +1,198 @@
+        const basePath = _Abs('/sub_entry_points/node_modules');
+        loadTestFiles([
+          ...createPackage(fs.resolve(basePath, ''), 'common'),
+          ...createPackage(fs.resolve(basePath, 'common'), 'http', ['@angular/common']),
```
This is the "targeted" entry-point finding, which means that you must pass a target entry-point where the processing begins. In this case the target actually is `common`, which in this scenario has no dependencies, leading to a only a single entry-point being found.

petebacondarwin replied:
I will make it more obvious, hopefully, by introducing a `targetPath` variable.

alxhub commented on packages/compiler-cli/src/ngtsc/file_system/src/cached_node_js_file_system.ts:
```diff
@@ -0,0 +1,72 @@
+ * Be aware that any changes to the file system from outside of this
+ * class could break the cache, leaving it with stale values.
+ */
+export class CachedNodeJSFileSystem extends NodeJSFileSystem {
```
I would rather we:

1) Implement `CachedFileSystem` as its own thing which delegates to another `FileSystem`, or
2) Roll the caching behavior directly into `NodeJSFileSystem`.

#1 I think is preferred as it makes testing the caching implementation very straightforward, and allows us to cache against other FS implementations at some point. There's no special reason that caching should be only a specialization on top of the node FS.

petebacondarwin replied:
OK, I'm going to make it a delegating system. I will not use a `Proxy` approach but hard code each `FileSystem` method to call the delegate as necessary. This provides a safe typing situation and makes it clear which methods are being customized.

petebacondarwin replied:
Done

alxhub commented:
The cache isn't used in `ngtsc`, correct?

My concern would be that for a watch-mode compilation, we don't have logic in place yet to invalidate the FS cache when things change on disk. And the FS is a long-lived instance, so that would cause major problems.

alxhub reviewed (approved):
Modulo comments.

petebacondarwin commented:
> The cache isn't used in ngtsc, correct?

@alxhub - that is correct. As you say, there is no ability to invalidate the cache in this implementation - although it would not be very difficult to add it - so it is only being used for the ngcc tool, which does not have a long running FS.

JoostK commented on packages/compiler-cli/src/ngtsc/file_system/src/cached_node_js_file_system.ts:
```diff
@@ -0,0 +1,114 @@
+ * Be aware that any changes to the file system from outside of this
+ * class could break the cache, leaving it with stale values.
+ */
+export class CachedNodeJSFileSystem implements FileSystem {
```
Should this now be named just `CachedFileSystem`?

petebacondarwin replied:
LOL! Yes!

JoostK commented on packages/compiler-cli/src/ngtsc/file_system/test/cached_node_js_file_system_spec.ts:
```diff
@@ -0,0 +1,255 @@
+  let xyzPath: AbsoluteFsPath;
+
+  beforeEach(() => {
+    delegate = new NodeJSFileSystem();
```
Should we use `MockFileSystem` here instead? Looks like this instance is spied on in its entirety so I'm not sure if it matters much but it would make more sense to me.

petebacondarwin replied:
`MockFileSystem` is abstract so then we would need to use one of the concrete ones any way. My thoughts went like this: Either we create a specific FS for this test case, which contains spies or we spy on an FS that we have already. Creating a specific one is quite a lot of unnecessary boilerplate. If using a concrete one which is spied upon, why use one concrete system over another? Why not just use the "real" one?

JoostK replied:
Ah, I missed the abstractness of `MockFileSystem`. In that case, using the real impl is fine 👍

gkalpak replied:
I am a little confused. Do we need the actual implementation? If not why not create a mock one with just spies (e.g. [jasmine.createSpyObj()][1])?

[1]: https://jasmine.github.io/api/3.4/jasmine.html#.createSpyObj

petebacondarwin replied:
Yes we can totally do that, but what is the benefit? Creating a mock one with spies requires implementation of the `FileSystem` interface, which I found was a lot of boilerplate. So it seemed simpler to just spy on the real FS instead.

gkalpak replied:
As discussed "offline", we need actual implementations for some of the methods (such as `dirname()` and `resolve()`), so it makes sense to use (and spy) an existing implementation.

gkalpak commented on packages/compiler-cli/ngcc/src/entry_point_finder/utils.ts:
```diff
@@ -0,0 +1,65 @@
+  if (pathMappings) {
+    const baseUrl = resolve(pathMappings.baseUrl);
+    Object.values(pathMappings.paths).forEach(paths => paths.forEach(path => {
+      basePaths.push(join(baseUrl, extractPathPrefix(path)));
```
As reported in #31424, I think we need to also take into account that `path` might point point to a file, not a directory or a glob pattern.
Also, since `path` is relative to the `baseUrl`, wouldn't it make more sense to use `resolve()` instead of `join()` here?

petebacondarwin replied:
Good call @gkalpak - I'll update the PR regarding #31424.

`join` is faster than `resolve` (marginally) and given that `baseUrl` has already been resolved there shouldn't be any issues with drive prefixes etc on Windows, so I think it is better use `join` here.

petebacondarwin replied:
Added a commit to fix this.

gkalpak replied:
Oh, I didn't realize `.join()` resolves `..` segments 😳
So, `.join()` is fine :smiley:

petebacondarwin committed:
fix(ivy): ngcc - handle pathMappings to files rather then directories

Paths can be mapped directly to files, which was not being taken
into account when computing `basePaths` for the `EntryPointFinder`s.

Now if a `pathMapping` pattern does not exist or is a file, then we try
the containing folder instead.

Fixes #31424

gkalpak commented on packages/compiler-cli/ngcc/src/entry_point_finder/utils.ts:
```diff
@@ -29,11 +29,17 @@ import {PathMappings} from '../utils';
     const baseUrl = resolve(pathMappings.baseUrl);
     Object.values(pathMappings.paths).forEach(paths => paths.forEach(path => {
-      basePaths.push(join(baseUrl, extractPathPrefix(path)));
+      // We only want base paths that exists and are not files
```
Typo: `exists` --> `exist`

gkalpak commented on packages/compiler-cli/ngcc/src/entry_point_finder/utils.ts:
```diff
@@ -29,11 +29,17 @@ import {PathMappings} from '../utils';
-      basePaths.push(join(baseUrl, extractPathPrefix(path)));
+      // We only want base paths that exists and are not files
+      let basePath = join(baseUrl, extractPathPrefix(path));
+      while (basePath !== baseUrl && !fs.exists(basePath) || fs.stat(basePath).isFile()) {
```
You probably meant `basePath !== baseUrl && (!fs.exists(basePath) || fs.stat(basePath).isFile())`. It shouldn't affect the final outcome here (since `baseUrl` should never be a file), but it means that `fs.stat()` could run unnecessarily.

petebacondarwin replied:
Oh! I always thought that `&&` and `||` had equal precedence. I see that is wrong. This is definitely not correct, in that case since `stat()` will throw if the path doesn't exist.

gkalpak reviewed (approved):
I think it would be useful to have a test with a path mapped to a parent directory (something like `paths: {'@foo': ['../foo']}`) to make sure `..` segments are resolved correctly.

Otherwise lgtm :+1:

petebacondarwin committed:
fixup! fix(ivy): ngcc - handle pathMappings to files rather then directories

petebacondarwin commented:
Added a test.

petebacondarwin commented:
Merge assistance - please run presubmit.

jasonaden commented:
[Presubmit](http://test/OCL:257006948:BASE:257006979:1562606807413:6550eed8)
